mautic/mautic
# Switch to tracking via JS

## Description
This PR adds the ability to track users / visitors via JS, rather than the tracking pixel. It still uses the tracking pixel (loaded via JS) for BC (will be removed in 2.0), but it's the first step along the road to allow us to update to fully JS based tracking. This also allows plugins to hook into the building of the js and append their custom js to it from within mautic. This has implications for full universal tag management ability within Mautic, much like Google's UTM (ok, maybe not THAT advanced).

This PR is required for the Web Push Channel to function.

## Steps to test

Merge the PR into your local testing, then clear your cache. You can then load `{your site}/mtc.js` and see the JS output. 

To test plugin integration, create an event listener that subscribes to the `CoreEvents::BUILD_MAUTIC_JS` event. From within your subscribed method, use `$event->appendJs($js, $section)`, where `$section` is the (optional) name / purpose of the JS you are adding, and `$js` is the actual javascript. See https://github.com/dongilbert/mautic/blob/56973d0ad6009679f1f77df635ff2bc61272ca74/app/bundles/CoreBundle/EventListener/BuildJsSubscriber.php#L36 for an example. 

The JS output here is automatically minified, unless your mautic install is in debug mode.

Finally, you are also able to re-order the loading priority of the JS that is added to the mtc.js by changing the priority within your subscribers `getSubscribedEvents()` method. By default, the tracking pixel js is loaded first, and has a top priority of 255.
